### ISSUE TITLE
Table & Multiline text support?

### ISSUE THREAD

**Issue by aaronjrich** [ttutisani/Xunit.Gherkin.Quick#20]
Hi, was wondering if tables or multi-line text is supported in this implementation?  My attempts to use them weren't working.

Thanks much.

**Comment by ttutisani** [comment]
You're right, tables are not supported. Not sure about the multiline. Do you mean DocString? if yes, that's also not supported.
Table support is planned and even was in progress. We may have to redo that. @csurfleet what's the status on that? if you're not available, let me know and I can take that over.
DocString - we will work on it too. If you have time, you can do a PR too. If not, then will be done in one of next releases. Good point.

**Comment by aaronjrich** [comment]
I took a stab at some basic support for these two so they will at least not fail when using them.  Examples of what I what thinking in the add numbers feature file.

**Comment by csurfleet** [comment]
@ttutisani where are we on this now? I can spend some working time on this, as we're using this for an official project now :D

**Comment by wcdeich4** [comment]
Hi.  Xunit.Gherkin.Quick is awesome!!!!!!!!  Really looking forward to full "Example table" support - that is one of the more fundamental features of Gherkin that software clients want.  And if you could support putting the example table in a separate file, like JBehave supports, that would be even more awesome!!!!! 

This is what I am talking about
@metaTag
Scenario Outline: 
Given .........
When ..........
Then..........
Examples: 
.\Data\Table.txt


And Table.text would contain:  
|x|y|z|
|1|2|3|
|2|4|6|

**Comment by ttutisani** [comment]
Currently I'm working on a codebase refactoring. Once done with that, it should be straightforward to add table support.

I'll keep the table file support as next step, after inline table works.

**Comment by ttutisani** [comment]
BTW, as I suggested to many people too many times, PR is welcome, if you want to add support for the table argument.
We had 2 PRs and both somehow didn't make it to the merge. I do review PRs and expect comments to be addressed - I hope this expectation is reasonable.

I'll account for and reconcile all changes into the refactoring that I'm working on: https://github.com/ttutisani/Xunit.Gherkin.Quick/issues/25

Another option is to just wait until I reach the point to add such support, which will be after I'm done with the refactoring.

**Comment by wcdeich4** [comment]
Well, if we are both lucky, maybe by the time you are done refactoring, I can find free time after work to make a pull request :-)

**Comment by ttutisani** [comment]
No worries :-) Refactoring seems to be halfway through.

**Comment by wcdeich4** [comment]
BTW, an example of "multiline" text can be found here:  https://github.com/serenity-bdd/serenity-jbehave/blob/master/src/test/resources/stories/BehaviorWithEmbeddedTables.story 

Basically, in addition to the Examples table at the bottom, Gherkin can support little per-line tables; but personally, I have never found it to be very useful.  Just putting all the required permutations into the examples table at the bottom can do whatever I need.

**Comment by wcdeich4** [comment]
That last example was JBehave style.  Here is one that is Cucumber style: 

https://github.com/serenity-bdd/serenity-cucumber/blob/master/src/test/resources/features/calculator/basic_arithmetic_with_tables.feature 

But again, just having 1 Examples: table at the bottom of each scenario is the most important part.

**Comment by timjroberts-inty** [comment]
Any timeframe on table support?

**Comment by ttutisani** [comment]
It seems the interest has increased for this feature.

I will be working on this with higher priority and should be out approximately by end of next week.

**Comment by timjroberts-inty** [comment]
Absolutely. Tables help reduce verbosity (Given...And...And...And...And).

This is a fantastic library by the way.

**Comment by ttutisani** [comment]
Thank you!

The work is in progress for DataTable argument ([branch compare](https://github.com/ttutisani/Xunit.Gherkin.Quick/compare/master...dataTableArgument?expand=1)).
Couple of days before merge, and then I'll release the nuget version.

Note that this will add support for bare minimum - DataTable type argument. You'd need to parse the cell values after you receive DataTable, because every cell has string value in it. `DataTable` type comes from Gherkin framework.

**Comment by ttutisani** [comment]
FYI. Code is merged now with DataTable argument support.
I will be adding a readme file now to explain the usage. Then I'll work on preparing the nuget version.

**Comment by timjroberts-inty** [comment]
Fantastic!

**Comment by wcdeich4** [comment]
The example at  https://github.com/ttutisani/Xunit.Gherkin.Quick/blob/master/datatable-argument.md   shows an "inline table" that applies to  just 1 line of Gherkin.   Does it also work for the common "Examples: " table that would be shared by all the lines of Gherkin?

If not, I may have some free time to write a pull request, not this weekend, but next weekend.

**Comment by ttutisani** [comment]
Short answer is no. These are 2 different things in Gherkin.

Specifically, what I've implemented is a `DataTable`, which comes as argument attached to the scenario step.
What you are asking for is `Examples` (as opposed to DataTable), which comes as examples attached to the scenario outline (as opposed to scenario step). This makes them very different - ScenarioOutline vs ScenarioStep (one is comparable to scenario, another is comparable to step in scenario).

Currently ScenarioOutline is not supported by this framework (maybe only by accidentally because it derives from ScenarioDefinition, just as Scenario does). However, ScenarioOutline's steps have placeholders, and without parsing parsing them, examples would not make sense. Unless you are misusing Gherkin and using Examples as you would use DataTable (you should not do that).

Anyway, if you have bandwidth to implement both ScenarioOutline and Examples support, please do so. If I reach this point before then, I will post here. If not, please post here when you start working on it, to avoid duplicate efforts.

Let me know if you have any questions about ScenarioOutline and Examples support.

**Comment by ttutisani** [comment]
@wcdeich4 let's discuss what you have in mind. I want to ensure we are on the same page.

Suppose we have this in Gherkin file:
```Gherkin
Scenario Outline: eating
  Given there are <start> cucumbers
  When I eat <eat> cucumbers
  Then I should have <left> cucumbers

  Examples:
    | start | eat | left |
    |    12 |   5 |    7 |
    |    20 |   5 |   15 |
```

I'd imagine this implemented this way:
- Just like implementing scenario steps, consumers would need to implement steps of the scenario outline. That give us 3 step methods.
- Each step method would need to accept parameters written in the Gherkin file. i.e. first step accepts "start", second step accepts "eat", third step accepts "left".
- Examples table is not provided to the step methods directly. Instead, framework infrastructure should parse the table, and execute each step by passing the arguments from the Examples table.

Is that what you had in mind?

Importantly, I don't see how this would be anyhow related to what your original comment was. No step would have direct access to the Examples table. It's just values that are being passed into the outline steps. Whole table is not accessible to the steps. That's why it's so different from DataTable, because with that, step has reference to DataTable and all its rows and cells.

Thoughts?

**Comment by wcdeich4** [comment]
Sorry not to respond sooner; work has been crazy the last couple of days.  Yes, what you described is what I had in mind.

When It comes to the data table being accessible to the steps, I think the data should "flow in the other direction" so to speak.  I'm not sure if I know how to word this properly, please so let me give an example.  Let's take the simplest possible feature & step files: 

Feature: AddTwoNumbers
Scenario: Add two numbers
	Given I chose 12 as first number
	And I chose 15 as second number
	When I press add
	Then the result should be 27 on the screen

[FeatureFile("./Story.feature")]
public sealed class AddTwoNumbers : Feature
{
    private int a,b,c;
    [Given(@"I chose (\d+) as first number")]
    public void I_chose_first_number(int firstNumber)
    {
        a = firstNumber;
    }
    [And(@"I chose (\d+) as second number")]
    public void I_chose_second_number(int secondNumber)
    {
        b = secondNumber;
    }
    [When(@"I press add")]
    public void I_press_add()
    {
        c = a + b;
    }
    [Then(@"the result should be (\d+) on the screen")]
    public void The_result_should_be_z_on_the_screen(int expectedResult)
    {
        Assert.Equal(a+b, c);
    }
}

OK, clearly the "FeatureFile" annotation class kicks off code that reads the .feature file & then decided which methods in the step class to call in which order & passes in the parameter(s) to the method as it calls it.

So, let's add an example table
Feature: AddTwoNumbers
Scenario: Add two numbers
	Given I chose <a> as first number
	And I chose <b> as second number
	When I press add
	Then the result should be <c> on the screen
Examples:
|a|b|c|
|1|2|3|
|4|5|9|

[FeatureFile("./Story.feature")]
public sealed class AddTwoNumbers : Feature
{
    private int a,b,c;
    [Given(@"I chose (\d+) as first number")]
    public void I_chose_first_number(int firstNumber)
    {
        a = firstNumber;
    }
    [And(@"I chose (\d+) as second number")]
    public void I_chose_second_number(int secondNumber)
    {
        b = secondNumber;
    }
    [When(@"I press add")]
    public void I_press_add()
    {
        c = a + b;
    }
    [Then(@"the result should be (\d+) on the screen")]
    public void The_result_should_be_z_on_the_screen(int expectedResult)
    {
        Assert.Equal(a+b, c);
    }
}

Now the behavior would be the same.  The FeatureFile annotation class still kicks off code that reads the .feature file & then decides which methods in the step class to call in which order & passes in the parameter(s) to the method as it calls it.  The only difference is, this time instead of just processing the statements of Gherkin from A, B, C, D... linearly, now you must loop through the rows of the examples table row=0  to  rows.Count-1 and fore each row in the table you execute the methods that correspond to Gherkin statements A, B, C, D....

**Comment by wcdeich4** [comment]
Please do not take this the wrong way, but I was  surprised in the last PR you opted for a step method signature of `public void method(int inputCount, DataTable dataTable)`.  There is never a need for the step method to own the data or control the data or even know what index it is at.  The BDD system decides which methods to to call, what data to pass as parameters & how many times to call them.  That way the same step method can be called by different Gherkin files, some of which have an in-line-multi-row-data, some of which have an examples table at the bottom & some of which have no rows, only data in the Gherkin statement itself.

**Comment by ttutisani** [comment]
About the Examples. You are right. This is exactly what I was suggesting.
Only small corrections:

Your example said `Scenario: Add two numbers`, but it should say `Scenario Outline: Add two numbers`. Because only scenario outline has Examples. Scenario does not have Examples table.

Your example showed scenario steps and their format. Instead, steps should have parameters enclosed in tag brackets, e.g. `Given I chose <a> as first number`.

And, step methods would need to say `[Given(@"I chose <a> as first number")]`. That will ensure that they match the step text.

But I think you assumed this all. So we must be on the same page.
I just wanted to set clear expectation. That's the goal to have it working this way.
But if you doubt about what I wrote, please let me know. Let's clarify before coding, to avoid waste of time and rewrites.

**Comment by ttutisani** [comment]
About your second comment for my last PR and this code `public void method(int inputCount, DataTable dataTable)`.

I didn't quite understand what you mean.
I see you are trying to mix DataTable with Examples. These work differently. Framework will not (and should not) call same step many times for each row of the DataTable. You are confusing DataTable with Examples. Please don't.
DataTable is just an argument, just like number or DocString.

Here is a quote from Gherkin official reference (https://docs.cucumber.io/gherkin/reference/) on the Cucumber repo:

> Data Tables are handy for passing a list of values to a step definition:
>
> Given the following users exist:
>  | name   | email              | twitter         |
>  | Aslak  | aslak@cucumber.io  | @aslak_hellesoy |
>  | Julien | julien@cucumber.io | @jbpros         |
>  | Matt   | matt@cucumber.io   | @mattwynne      |
>
>Just like Doc Strings, Data Tables will be passed to the step definition as the last argument.

This is a quote. Not my definition.
Let me repeat from the quote:

Just like Doc Strings, Data Tables will be passed to the **_step definition_** as the last **_argument_**.

I hope this clarifies. DataTable is an argument. And it's a Step argument. DataTable should be passed to the step. Step should not receive datatable row values just so.
Example works differently though.

If you have confusions or questions, I'll be happy to discuss further.

**Comment by wcdeich4** [comment]
Hi.  I did  not have free time this weekend after all.  I'm really sorry.  Thank you for explaining how DataTables are implemented differently according to the Cucumber Gherkin standard; actually I always found Example tables to be so much more useful, that I never really used DataTables & just assumed they worked the same.  

About "[Given(@"I chose <a> as first number")]" --- that is exactly how it is done in JBehave.  In Cucumber they use a regular expression, which makes it a little more flexible --- in particular, the variable does not need to be named "a"
